### PR TITLE
Campaign admin edit/add view

### PR DESCRIPTION
**Pull request by flashbackzoo**
This adds functionality for creating and editing Campaigns in the Campaign admin section.

# Creating

campaign-add-index-before

campaign-add-add

campaign-add-after

## Editing

campaign-admin-edit-1

Clicking the 'cog' icon takes you to...

campaign-admin-edit-2

### CODE REVIEW

File: admin/client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 1]
```diff
@@ -255,7 +255,12 @@ class CampaignAdmin extends SilverStripeComponent {
   }
 
   addCampaign() {
-    // Add campaign
+    const path = this.props.sectionConfig.campaignViewRoute
+      .replace(/:type\?/, 'set')
+      .replace(/:id\?/, 0)
```
Comment (chillu) [inline admin/client/src/containers/CampaignAdmin/CampaignAdmin.js]: Adding a `0` for a creation URL is pretty ugly. Also, the routing doesn't work when going to `/admin/campaigns/set/0/create` directly.

File: admin/code/CampaignAdmin.php [hunk 2]
```diff
@@ -486,7 +486,8 @@ public function getCreateEditForm() {
 			'CreateEditForm',
 			ChangeSet::singleton()->getCMSFields(),
 			FieldList::create(
-				FormAction::create('save', 'Save')
+				FormAction::create('save', 'Save'),
+				FormAction::create('cancel', 'Cancel')
```
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Needs i18n

File: admin/code/CampaignAdmin.php [hunk 3]
```diff
@@ -464,7 +458,8 @@ public function getDetailEditForm($id) {
 			'DetailEditForm',
 			$fields,
 			FieldList::create(
-				FormAction::create('save', 'Save')
+				FormAction::create('save', 'Save'),
+				FormAction::create('cancel', 'Cancel')
```
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Needs i18n

File: model/versioning/ChangeSet.php [hunk 4]
```diff
@@ -346,10 +346,12 @@ public function can($perm, $member = null, $context = array()) {
 
 	public function getCMSFields() {
 		$fields = new FieldList();
-		$fields->merge([
-			TextField::create('Name'),
-			ReadonlyField::create('State')
-		]);
+		$fields->push(TextField::create('Name'));
```
Comment (chillu) [inline model/versioning/ChangeSet.php]: Needs i18n

File: admin/client/src/containers/CampaignAdmin/CampaignAdmin.js [hunk 5]
```diff
@@ -158,6 +162,53 @@ class CampaignAdmin extends SilverStripeComponent {
...
+   * @param {Object} Component - Component constructor.
+   * @param {Object} props - Props passed from FormBuilder.
+   * @return {Object} - Instanciated React component
+   */
+  campaignCreationCreateFn(Component, props) {
+    if (props.name === 'action_save') {
+      const extendedProps = Object.assign({}, props, {
+        type: 'submit',
+        label: props.title,
+        icon: 'save',
```
Comment (chillu) [inline admin/client/src/containers/CampaignAdmin/CampaignAdmin.js]: Should be styled green, same as edit view

File: admin/code/CampaignAdmin.php [hunk 6]
```diff
@@ -490,6 +476,22 @@ public function getDetailEditForm($id) {
 	}
 
 	/**
+	 * @todo Use GridFieldDetailForm once it can handle structured data and form schemas
+	 *
+	 * @return Form
+	 */
+	public function getCreateEditForm() {
```
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Shouldn't this just be called `CreateForm`?

File: admin/code/CampaignAdmin.php [hunk 7]
```diff
@@ -490,6 +476,22 @@ public function getDetailEditForm($id) {
...
+		return Form::create(
+			$this,
+			'CreateEditForm',
+			ChangeSet::singleton()->getCMSFields(),
+			FieldList::create(
+				FormAction::create('save', 'Save')
```
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Needs i18n

File: admin/code/CampaignAdmin.php (hunk 6 above)
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Has been refactored since, now just using `DetailForm`

File: model/versioning/ChangeSet.php (hunk 4 above)
Comment (chillu) [inline model/versioning/ChangeSet.php]: Fixed

File: admin/code/CampaignAdmin.php (hunk 3 above)
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Fixed

File: admin/code/CampaignAdmin.php (hunk 2 above)
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Fixed

File: admin/code/CampaignAdmin.php (hunk 7 above)
Comment (chillu) [inline admin/code/CampaignAdmin.php]: Fixed

File: admin/client/src/containers/CampaignAdmin/CampaignAdmin.js (hunk 1 above)
Comment (flashbackzoo) [inline admin/client/src/containers/CampaignAdmin/CampaignAdmin.js]: Yup - fixing now

**Comment by tractorcow** [comment]
I get these errors running locally

```
There were 3 failures:

1) TextareaFieldTest::testReadonlyDisplayUnicodes
Failed asserting that '
	<span id="Test" class="readonly textarea">
		
	</span>

' contains "These are some unicodes: äöü".

/Users/dmooyman/Sites/ss40test/framework/dev/SapphireTest.php:546
/Users/dmooyman/Sites/ss40test/framework/tests/forms/TextareaFieldTest.php:24
/Users/dmooyman/Sites/ss40test/vendor/phpunit/phpunit/phpunit:47

2) TextareaFieldTest::testReadonlyDisplaySepcialHTML
Failed asserting that '
	<span id="Test" class="readonly textarea">
		
	</span>

' contains "These are some special &lt;html&gt; chars including &#039;single&#039; &amp; &quot;double&quot; quotations".

/Users/dmooyman/Sites/ss40test/framework/dev/SapphireTest.php:546
/Users/dmooyman/Sites/ss40test/framework/tests/forms/TextareaFieldTest.php:36
/Users/dmooyman/Sites/ss40test/vendor/phpunit/phpunit/phpunit:47

3) GridFieldDetailFormTest::testViewForm
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Jane'
+''

/Users/dmooyman/Sites/ss40test/framework/tests/forms/gridfield/GridFieldDetailFormTest.php:113
/Users/dmooyman/Sites/ss40test/vendor/phpunit/phpunit/phpunit:47

FAILURES!
Tests: 1784, Assertions: 7680, Failures: 3, Skipped: 30, Incomplete: 5.
```

**Pull request merged**